### PR TITLE
Unbuffered Stream Reader

### PR DESCRIPTION
**Pull request by mbaker3**
Adds `UnbufferedStreamReader`. A stream reader that does not buffer data from the backing BaseStream.
This implementation is less performant than `StreamReader` but doesn't read ahead on the underlying stream ensuring the BaseStream's position to remains lockstep with the characters read out of `UnbufferedStreamReader`.

This stream is useful when reading files with mixed encoding or where the BaseStream needs to be advanced the minimum amount required.

### What is the current behaviour?
`StreamReader` fills a buffer of characters when responding to read requests. The characters buffer is filled by repeatedly reading blocks of its `BaseStream` and converting them into characters until the character buffer is full.

This means that the `BaseStream` is advanced beyond the point that the `StreamReader` has been read.

### What is the new behaviour?
`UnbufferedStreamReader` effectively functions the same as `StreamReader` with a byte and character buffer size set to 1. The `UnbufferedStreamReader` reads one byte at a time out of `BaseStream` until it has enough to resolve a full character. That one character is then evaluated by the read method to decide whether to continue reading or return.

**Note:** The implementation does not currently cover the complete features set of `StreamReader`. New features will be developed as needed. This effort is captured in #87 

#### Why don't you just calculate actual the `BaseStream` position based on `StreamReader`'s unused buffer contents?
TL;DR
 - Invalid bytes are replaced with a pre-defined fallback character which may not equal the byte size of the invalid bytes.
 - The reflection relies heavily on Microsoft not changing the way `StreamReader` is implemented
 - It does not work for variable-byte encodings (Ex: UTF-7)

I went down that path. It's not possible. [This StackOverflow post](https://stackoverflow.com/questions/5404267/c-sharp-streamreader-and-seeking/17457085#17457085) covers the issue in detail. Read the linked post + all comments to understand.

If you really want to dig into this read through `StreamReader`'s source: https://referencesource.microsoft.com/#mscorlib/system/io/streamreader.cs,b5fe1efcec14de32

### What issues does this resolve?
None, This class allows developers to continue reading a `BaseStream` beyond the lifecycle of the `UnbufferedStreamReader`.

### What PRs does this depend on?
 - None

### Does this introduce a breaking change?
 - [ ] Yes
 - [x] No

### CODE REVIEW

File: IO/UnbufferedStreamReader.cs [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+namespace Anvil.CSharp.IO
+{
+    /// <summary>
+    /// A stream reader that does not use a buffer.
+    /// <see cref="StreamReader"/> is more performant but advances its <see cref="BaseStream"/> past its own position an indeterminate amount.
+    /// This stream is useful when reading files with mixed encoding or where the <see cref="BaseStream"/> needs to be advanced the minimum amount required.
+    /// </summary>
+    /// <remarks>
+    /// This reader is not currently at feature parity with <see cref="StreamReader"/>. Features will be developed as needed.
+    /// See: #87
```
Comment (jkeon) [inline IO/UnbufferedStreamReader.cs:15]: Should we link that SO post in here too? Useful background info.

File: IO/UnbufferedStreamReader.cs [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+        /// </summary>
+        public Encoding CurrentEncoding { get; }
+
+        private readonly bool m_LeaveBaseStreamOpen;
+        private readonly Decoder m_Decoder;
+        private readonly StringBuilder m_StringBuilder;
+
+        // Used for APIs that require a buffer. All single length;
+        private readonly byte[] m_SingleByteBuffer = new byte[1];
+        private readonly char[] m_charBuffer = new char[1];
```
Comment (jkeon) [inline IO/UnbufferedStreamReader.cs:34]:
```suggestion
        private readonly char[] m_CharBuffer = new char[1];
```

File: IO/UnbufferedStreamReader.cs [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+                    return default;
+                }
+
+                decodedCharCount = m_Decoder.GetChars(m_SingleByteBuffer, 0, 1, m_charBuffer, 0);
+            } while (decodedCharCount < 1);
+
+            return m_charBuffer[0];
+        }
+
+        private string RenderAndClearStringBuilder()
```
Comment (jkeon) [inline IO/UnbufferedStreamReader.cs:180]: Seems odd to call it `Render`?  Maybe `ResolveAndClearStringBuilder`? `ExtractAndClear`?

**Review by jkeon — CHANGES_REQUESTED** [review]
Minor comments, otherwise looks good

File: IO/UnbufferedStreamReader.cs (hunk 3 above)
Comment (mbaker3) [reply IO/UnbufferedStreamReader.cs:180]: I mean, you are rendering a string from the builder's component parts but sure I've rename it to resolve.

**Comment by mbaker3** [comment]
Feedback addressed

File: IO/UnbufferedStreamReader.cs (hunk 3 above)
Comment (jkeon) [reply IO/UnbufferedStreamReader.cs]: Not to the screen man.... not to the screen.

**Review by jkeon — APPROVED** [review]
